src/components/tests/SourcesTree.spec.js

```diff
@@ -85,4 +86,622 @@ describe("SourcesTree", () => {
+          ...props
+        });
+        expect(component).toMatchSnapshot();
+        expect(props.selectLocation).toHaveBeenCalledWith({
+          sourceId: "server1.conn13.child1/41"
+        });
+      });
+    });
+
+    describe("updates highlighted items", () => {
+      it("updates highlightItems if selectedSource changes", async () => {
+        props.selectedSource = sourceThree;
+        await component.setProps({
+          ...props
+        });
+        expect(component).toMatchSnapshot();
+      });
+    });
+  });
+
+  describe("focusItem", () => {
+    it("update the focused item", async () => {
+      const { component } = render({});
+
+      expect(component.state("focusedItem")).toEqual(null);
+
+      await component.instance().focusItem(itemOne);
+
+      expect(component.state("focusedItem")).toEqual(itemOne);
+    });
+  });
+
+  describe("with custom root", () => {
+    const { component, props } = render({
+      projectRoot: "mdn",
+      clearProjectDirectoryRoot: jest.fn()
+    });
+
+    afterEach(() => {
+      props.clearProjectDirectoryRoot.mockClear();
+    });
+
+    it("renders custom root source list", async () => {
+      expect(component).toMatchSnapshot();
+    });
+
+    it("renders empty custom root source list", async () => {
+      const emptyComponent = render({
+        projectRoot: "custom",
+        sources: I.Map()
+      }).component;
+      expect(emptyComponent).toMatchSnapshot();
+    });
+
+    it("calls clearProjectDirectoryRoot on click", async () => {
+      component.find(".sources-clear-root").simulate("click");
+      expect(props.clearProjectDirectoryRoot).toHaveBeenCalled();
+    });
+  });
+
+  describe("onContextMenu", () => {
+    let component;
+    let props;
+
+    const event = {
+      preventDefault: jest.fn(),
+      stopPropagation: jest.fn()
+    };
+
+    const directory = {
+      contents: new Map(),
+      name: "folder",
+      path: "folder/"
+    };
+
+    const file = {
+      contents: new Map(),
+      name: "file",
+      path: "folder/file.js"
+    };
+
+    const rootDirectory = {
+      contents: new Map(),
+      name: "root",
+      path: "root/"
+    };
+
+    beforeEach(() => {
+      const rendered = render({
+        clearProjectDirectoryRoot: jest.fn(),
+        setProjectDirectoryRoot: jest.fn(),
+        projectRoot: "root/"
+      });
+      component = rendered.component;
+      props = rendered.props;
+    });
+
+    afterEach(() => {
+      showMenu.mockClear();
+      event.preventDefault.mockClear();
+      event.stopPropagation.mockClear();
+    });
+
+    it("shows context menu on directory to set as root", async () => {
+      const menuOptions = [
+        {
+          accesskey: "r",
+          click: expect.any(Function),
+          disabled: false,
+          id: "node-set-directory-root",
+          label: "Set directory root"
+        }
+      ];
+      await component.instance().onContextMenu(event, directory);
```

Comment:
we use instance quite a bit, maybe it should be returned from the render function?

Reply:
not sure it's better to do this. I think it will make it more messy now that I have co-located the different components and would need a different instance for each component.